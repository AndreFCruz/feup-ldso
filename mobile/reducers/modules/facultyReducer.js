// Actions
// Set Faculty
const SET_FACULTY = 'mobile/faculty/SET_FACULTY';
const CLEAR_FACULTY = 'mobile/faculty/CLEAR_FACULTY';
// Get Faculty Introduction
const GET_STATS = 'mobile/faculty/GET_STATS';
const GET_STATS_SUCCESS = 'mobile/faculty/GET_STATS_SUCCESS';
const GET_STATS_FAIL = 'mobile/faculty/GET_STATS_FAIL';
// Get Faculty Social Projects
const GET_SOCIAL_PROJECTS = 'mobile/faculty/GET_SOCIAL_PROJECTS';
const GET_SOCIAL_PROJECTS_SUCCESS = 'mobile/faculty/GET_SOCIAL_PROJECTS_SUCCESS';
const GET_SOCIAL_PROJECTS_FAIL = 'mobile/faculty/GET_SOCIAL_PROJECTS_FAIL';
// Get Faculty Social Project by ID
const GET_SOCIAL_PROJECT_BY_ID = 'mobile/faculty/GET_SOCIAL_PROJECT_BY_ID';
const GET_SOCIAL_PROJECT_BY_ID_SUCCESS = 'mobile/faculty/GET_SOCIAL_PROJECT_BY_ID_SUCCESS';
const GET_SOCIAL_PROJECT_BY_ID_FAIL = 'mobile/faculty/GET_SOCIAL_PROJECT_BY_ID_FAIL';
// Get Research Centres
const GET_RESEARCH_CENTRES = 'mobile/faculty/GET_RESEARCH_CENTRES';
const GET_RESEARCH_CENTRES_SUCCESS = 'mobile/faculty/GET_RESEARCH_CENTRES_SUCCESS';
const GET_RESEARCH_CENTRES_FAIL = 'mobile/faculty/GET_RESEARCH_CENTRES_FAIL';
// Get Research Centre by ID
const GET_RESEARCH_CENTRE_BY_ID = 'mobile/faculty/GET_RESEARCH_CENTRE_BY_ID';
const GET_RESEARCH_CENTRE_BY_ID_SUCCESS = 'mobile/faculty/GET_RESEARCH_CENTRE_BY_ID_SUCCESS';
const GET_RESEARCH_CENTRE_BY_ID_FAIL = 'mobile/faculty/GET_RESEARCH_CENTRE_BY_ID_FAIL';
// Get Faculty Future Prospects
const GET_FUTURE_PROSPECTS = 'mobile/faculty/GET_FUTURE_PROSPECTS';
const GET_FUTURE_PROSPECTS_SUCCESS = 'mobile/faculty/GET_FUTURE_PROSPECTS_SUCCESS';
const GET_FUTURE_PROSPECTS_FAIL = 'mobile/faculty/GET_FUTURE_PROSPECTS_FAIL';
// Get Faculty Localization
const GET_LOCALIZATION = 'mobile/faculty/GET_LOCALIZATION';
const GET_LOCALIZATION_SUCCESS = 'mobile/faculty/GET_LOCALIZATION_SUCCESS';
const GET_LOCALIZATION_FAIL = 'mobile/faculty/GET_LOCALIZATION_FAIL';
// Get Faculty Videos
const GET_VIDEOS = 'mobile/faculty/GET_VIDEOS';
const GET_VIDEOS_SUCCESS = 'mobile/faculty/GET_VIDEOS_SUCCESS';
const GET_VIDEOS_FAIL = 'mobile/faculty/GET_VIDEOS_FAIL';

const initialState = {
    loading: false,
    name: '',
    stats: {},
    socialProjects: [],
    currSocialProject: null,
    researchCentres: [],
    currResearchCentre: null,
    futureProspects: {},
    localization: null,
    videos: []
};

export default function reducer(state = initialState, action) {
    switch (action.type) {
    case SET_FACULTY:
        return { ...state,
            name: action.name
        };
    case CLEAR_FACULTY:
        return initialState;

    case GET_STATS:
    case GET_SOCIAL_PROJECTS:
    case GET_SOCIAL_PROJECT_BY_ID:
    case GET_RESEARCH_CENTRES:
    case GET_RESEARCH_CENTRE_BY_ID:
    case GET_FUTURE_PROSPECTS:
    case GET_LOCALIZATION:
    case GET_VIDEOS:
        return { ...state,
            loading: true
        };

    case GET_STATS_SUCCESS:
        return { ...state,
            loading: false,
            stats: action.payload.data[0]
        };
    case GET_SOCIAL_PROJECTS_SUCCESS:
        return { ...state,
            loading: false,
            socialProjects: action.payload.data
        };
    case GET_SOCIAL_PROJECT_BY_ID_SUCCESS:
        return { ...state,
            loading: false,
            currSocialProject: action.payload.data[0]
        };
    case GET_RESEARCH_CENTRES_SUCCESS:
        return { ...state,
            loading: false,
            researchCentres: action.payload.data
        };
    case GET_RESEARCH_CENTRE_BY_ID_SUCCESS:
        return { ...state,
            loading: false,
            currResearchCentre: action.payload.data[0]
        };
    case GET_FUTURE_PROSPECTS_SUCCESS:
        return { ...state,
            loading: false,
            futureProspects: action.payload.data.future_prospects
        };
    case GET_LOCALIZATION_SUCCESS:
        return { ...state,
            loading: false,
            localization: action.payload.data
        };
    case GET_VIDEOS_SUCCESS:
        return { ...state,
            loading: false,
            videos: action.payload.data.videos
        };

    case GET_STATS_FAIL:
    case GET_SOCIAL_PROJECTS_FAIL:
    case GET_SOCIAL_PROJECT_BY_ID_FAIL:
    case GET_RESEARCH_CENTRES_FAIL:
    case GET_RESEARCH_CENTRE_BY_ID_FAIL:
    case GET_FUTURE_PROSPECTS_FAIL:
    case GET_LOCALIZATION_FAIL:
    case GET_VIDEOS_FAIL:
        return {
            ...state,
            loading: false,
            error: 'Error while fetching faculty data'
        };
    default:
        return state;
    }
}

export function setFaculty(name) {
    return {
        type: SET_FACULTY,
        name: name
    };
}

export function clearFaculty() {
    return {
        type: CLEAR_FACULTY
    };
}

export function getStats(language, faculty) {
    return {
        type: GET_STATS,
        payload: {
            request: {
                url: `/faculties/${language}/${faculty}/statistics`
            }
        }
    };
}

export function getSocialProjects(language, faculty) {
    return {
        type: GET_SOCIAL_PROJECTS,
        payload: {
            request: {
                url: `/faculties/${language}/${faculty}/social-projects-short`
            }
        }
    };
}

export function getSocialProjectDetails(language, faculty, id) {
    return {
        type: GET_SOCIAL_PROJECT_BY_ID,
        payload: {
            request: {
                url: `/faculties/${language}/${faculty}/social-projects?filter[where][id]=${id}`
            }
        }
    };
}

export function getResearchCentres(language, faculty) {
    return {
        type: GET_RESEARCH_CENTRES,
        payload: {
            request: {
                url: `/faculties/${language}/${faculty}/research-centers-short`
            }
        }
    };
}

export function getResearchCentreDetails(language, faculty, id) {
    return {
        type: GET_RESEARCH_CENTRE_BY_ID,
        payload: {
            request: {
                url: `/faculties/${language}/${faculty}/research-centers?id=${id}`
            }
        }
    };
}

export function getFutureProspects(language, faculty) {
    return {
        type: GET_FUTURE_PROSPECTS,
        payload: {
            request: {
                url: `/faculties/${language}/${faculty}/future`
            }
        }
    };
}

export function getLocalization(faculty) {
    return {
        type: GET_LOCALIZATION,
        payload: {
            request: {
                url: `/${faculty}`
            }
        }
    };
}

export function getVideos(faculty) {
    return {
        type: GET_VIDEOS,
        payload: {
            request: {
                url: `/faculties/en/${faculty}/videos`
            }
        }
    };
}