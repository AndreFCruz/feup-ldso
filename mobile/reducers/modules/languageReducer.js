// Actions
// Toggle language between english and portuguese
const TOGGLE_LANGUAGE = 'mobile/language/TOGGLE_LANGUAGE';

const initialState = {
  selection: 'en'
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case TOGGLE_LANGUAGE:
      return {
        selection: state.selection == 'en'? 'pt' : 'en'
      };
    default:
      return state;
  }
}

export function toggleLanguage() {
  return {
    type: TOGGLE_LANGUAGE
  };
}