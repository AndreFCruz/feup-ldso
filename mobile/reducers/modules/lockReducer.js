// Actions
// Lock / Unlock Screen
const TOGGLE_LOCK = 'mobile/lock/TOGGLE_LOCK';

const initialState = {
    locked: false
};

// Draft reducer for lock logic - TODO
export default function reducer(state = initialState, action) {
    switch (action.type) {
    case TOGGLE_LOCK:
        return {
        // TODO será mais complexo do que isto, pois depende do boolean recebido pelo pedido a API (true se autenticar, falso otherwise)
            locked: !state.locked
        };
    default:
        return state;
    }
}

// Post request - user e password passados no objeto. Password já hashed. cofnrme resposta fazer
export function toggleScreen() {
    return {
        type: TOGGLE_LOCK
    
    };
}