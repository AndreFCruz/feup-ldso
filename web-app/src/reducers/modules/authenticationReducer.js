import { NotificationManager } from 'react-notifications';
import {
    LOGIN,
    LOGIN_SUCCESS,
    LOGIN_FAIL,
    LOGOUT
} from '../../actions/authenticationActions';
import {
    EDIT_PROJECT_FAIL,
    DELETE_PROJECT_FAIL,
    ADD_PROJECT_FAIL
} from '../../actions/projectsActions';
import {
    EDIT_PROSPECTS_FAIL
} from '../../actions/prospectsActions';
import {
    EDIT_VIDEO_FAIL
} from '../../actions/videosActions';

const initialState = {
    loading: false,
    authenticated: localStorage.getItem('iupToken') !== null
};

export default function reducer(state = initialState, action) {
    switch (action.type) {
    case LOGIN:
        return {
            ...state,
            loading: true
        };
    case LOGIN_SUCCESS: {
        return login(state, action.payload.data);
    }
    case LOGIN_FAIL:
        return {
            ...state,
            loading: false,
            error: action.payload.data.error
        };
    case LOGOUT: {
        return logout(state);
    }

    case EDIT_PROSPECTS_FAIL:
    case EDIT_VIDEO_FAIL:
    case ADD_PROJECT_FAIL:
    case DELETE_PROJECT_FAIL:
    case EDIT_PROJECT_FAIL: {
        let auth = true;
        if (action.payload.error.status === 401) {
            localStorage.removeItem('iupToken');
            auth = false;
        }

        return { ...state,
            authenticated: auth
        };
    }
    default:
        return state;
    }
}

const login = (state, response) => {
    const { token, error } = response;
    let authenticated = false;

    if (token === null && error !== '') {
        if (!global.__TEST__) NotificationManager.error(error);
    }
    else {
        localStorage.setItem('iupToken', token);
        if (!global.__TEST__) NotificationManager.success('Welcome back!');
        authenticated = true;
    }

    return {
        ...state,
        loading: false,
        authenticated: authenticated
    };
};

const logout = (state) => {
    localStorage.removeItem('iupToken');

    return {
        ...state,
        authenticated: false
    };
};