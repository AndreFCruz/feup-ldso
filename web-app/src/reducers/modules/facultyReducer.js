import {
    SET_FACULTY
} from '../../actions/facultyActions';

const initialState = {
    name: null
};

export default function reducer(state = initialState, action) {
    switch (action.type) {
    case SET_FACULTY:
        return { ...state,
            name: action.name
        };
    default:
        return state;
    }
}