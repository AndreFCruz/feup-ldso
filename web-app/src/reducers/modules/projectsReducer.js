import { NotificationManager } from 'react-notifications';
import {
    GET_PROJECTS_EN,
    GET_PROJECTS_EN_SUCCESS,
    GET_PROJECTS_EN_FAIL,
    GET_PROJECTS_PT,
    GET_PROJECTS_PT_SUCCESS,
    GET_PROJECTS_PT_FAIL,
    GET_PROJECT_BY_ID,
    ADD_PROJECT,
    EDIT_PROJECT,
    DELETE_PROJECT,
    GET_PROJECT_BY_ID_FAIL,
    ADD_PROJECT_FAIL,
    EDIT_PROJECT_FAIL,
    DELETE_PROJECT_FAIL,
    GET_PROJECT_BY_ID_SUCCESS,
    ADD_PROJECT_SUCCESS,
    EDIT_PROJECT_SUCCESS,
    DELETE_PROJECT_SUCCESS,
    RESET_PROJECTS,
} from '../../actions/projectsActions';

const initialState = {
    loading: false,
    loadingAction: false,
    projectsEN: [],
    projectsPT: [],
    currProjEN: null,
    currProjPT: null
};

export default function reducer(state = initialState, action) {
    switch (action.type) {
    case GET_PROJECTS_EN:
    case GET_PROJECTS_PT:
    case GET_PROJECT_BY_ID:
        return { ...state,
            loading: true
        };
    case ADD_PROJECT:
    case EDIT_PROJECT:
    case DELETE_PROJECT:
        return { ...state,
            loadingAction: true
        };
    case RESET_PROJECTS:
        return initialState;

    case GET_PROJECTS_EN_SUCCESS:
        return { ...state,
            loading: false,
            projectsEN: action.payload.data
        };
    case GET_PROJECTS_PT_SUCCESS:
        return { ...state,
            loading: false,
            projectsPT: action.payload.data
        };
    case GET_PROJECT_BY_ID_SUCCESS: {
        const projectByID = action.payload.data[0];
        return { ...state,
            loading: false,
            currProjEN: projectByID.language === 'en'? projectByID : state.currProjEN,
            currProjPT: projectByID.language === 'pt'? projectByID : state.currProjPT
        };
    }
    case ADD_PROJECT_SUCCESS: {
        return addProjects(state, action.payload.data);
    }
    case EDIT_PROJECT_SUCCESS: {
        return editProject(state, action.payload.data);
    }
    case DELETE_PROJECT_SUCCESS: {
        return deleteProject(state);
    }

    case GET_PROJECTS_EN_FAIL:
    case GET_PROJECTS_PT_FAIL:
    case GET_PROJECT_BY_ID_FAIL:
        return { ...state,
            loading: false,
            error: 'Error while fetching projects data'
        };
    case ADD_PROJECT_FAIL:
        if(!global.__TEST__) NotificationManager.error('Failed to add info');
        /* falls through */
    case EDIT_PROJECT_FAIL:
    case DELETE_PROJECT_FAIL:
        return { ...state,
            loadingAction: false,
            error: 'Error while executing action on project'
        };
    default:
        return state;
    }
}

const editProject = (state, editedProject) => {
    // Find project's index
    const editedIdx = state.projectsEN.findIndex(
        el => state.currProjEN.id === el.id
    );
    
    if(!global.__TEST__) NotificationManager.success('Successfully edited project!');
    if (editedProject.language === 'en') {
        let prevProps = [...state.projectsEN];
        prevProps[editedIdx] = editedProject;

        return {
            ...state,
            loadingAction: false,
            projectsEN: prevProps,
            currProjEN: editedProject
        };
    } else {
        let prevProps = [...state.projectsPT];
        prevProps[editedIdx] = editedProject;

        return {
            ...state,
            loadingAction: false,
            projectsPT: prevProps,
            currProjPT: editedProject
        };
    }
};

const addProjects = (state, newProjects) => {
    if(!global.__TEST__) NotificationManager.success('Successfully added new project!');
    return { ...state,
        loadingAction: false,
        projectsEN: [newProjects[0], ...state.projectsEN],
        projectsPT: [newProjects[1], ...state.projectsPT],
        currProjEN: newProjects[0],
        currProjPT: newProjects[1]
    };
};

const deleteProject = (state) => {
    let enIdx = state.projectsEN.findIndex((el) => el.id === state.currProjEN.id);
    
    if (enIdx >= 0) {
        let newProjectsEN = [...state.projectsEN];
        newProjectsEN.splice( enIdx, 1 );

        return { ...state,
            loadingAction: false,
            projectsEN: newProjectsEN,
            currProjEN: null
        };
    } else {
        let ptIdx = state.projectsPT.findIndex((el) => el.id === state.currProjPT.id);
        let newProjectsPT = [...state.projectsPT];
        newProjectsPT.splice( ptIdx, 1 );

        return { ...state,
            loadingAction: false,
            projectsPT: newProjectsPT,
            currProjPT: null
        };
    }
};