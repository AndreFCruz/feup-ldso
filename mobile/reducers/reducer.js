import { combineReducers } from 'redux';

import facultyReducer from './modules/facultyReducer';
import languageReducer from './modules/languageReducer';
import lockReducer from './modules/lockReducer';

const rootReducer =  combineReducers({
  faculty: facultyReducer,
  language: languageReducer,
  lock: lockReducer
});

export default rootReducer;
