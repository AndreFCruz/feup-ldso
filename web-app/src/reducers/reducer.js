import { combineReducers } from 'redux';

import facultyReducer from './modules/facultyReducer';
import statsReducer from './modules/statsReducer';
import projectsReducer from './modules/projectsReducer';
import videosReducer from './modules/videosReducer';

const rootReducer =  combineReducers({
    faculty: facultyReducer,
    stats: statsReducer,
    socialProjects: projectsReducer,
    videos: videosReducer
});

export default rootReducer;
