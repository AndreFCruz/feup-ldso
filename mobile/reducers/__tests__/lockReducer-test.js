import { toggleScreen } from '../modules/lockReducer';
import reducer from '../modules/lockReducer';

describe('Testing lock actions', () => {

    it('creates a TOGGLE_LOCK Action', () => {
        expect(toggleScreen()).toMatchSnapshot();
    });

});

describe('Testing if lock reducer', () => {

    it('changes locked state to true', () => {

        const initialState = {
            locked: false
        };

        expect(reducer(initialState, toggleScreen())).toMatchSnapshot();
    });

    it('changes locked state to false', () => {

        const initialState = {
            locked: true
        };

        expect(reducer(initialState, toggleScreen())).toMatchSnapshot();
    });

});

