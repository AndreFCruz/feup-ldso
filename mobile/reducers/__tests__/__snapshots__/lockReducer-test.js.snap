// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing if lock reducer changes locked state to false 1`] = `
Object {
  "locked": false,
}
`;

exports[`Testing if lock reducer changes locked state to true 1`] = `
Object {
  "locked": true,
}
`;

exports[`Testing lock actions creates a TOGGLE_LOCK Action 1`] = `
Object {
  "type": "mobile/lock/TOGGLE_LOCK",
}
`;
